### PR TITLE
[jormungandr] implement access_points in transfer path

### PR DESCRIPTION
**Pull request by xlqian**
With this PR. transfer paths are allowed between surface transport vehicles (`tramway`, `bus` etc...) and underground transport vehicles (`metro` etc...) **_via_** access points. 

Note that : 
* no transfer path will be calculated if access points are not available for the underlying underground transport vehicles. 
* no transfer path will be calculated if both stop points of the transfer are underground. (this case may be handled later)

How the access point (if multiple access points are present for the underlying stop point) is picked in a nutshell:

Let's say we need to calculate the path **from**  `stop_point:Tramway` **to** `stop_point: Metro`, given that  `stop_point: Metro` has 3 access points around it. The first step is to choose the eligible access points by requesting the `georef` all the access points of `stop_point:Metro` and filtering out those are `is_entrance == False`. Then we compute a matrix of `1xN` (reciprocally `Nx1` when transferring from `stop_point:Metro` to `stop_point:Tramway`). By adding up the `traversal_time` of each access points to the street network duration, we pick the most rapid access point to get to  `stop_point:Metro`. Then we launch the real street network computation between `stop_point:Tramway` and the chosen access point.

In the following example, the blue line (underground transport) has multiple access points,
image

but we finally chose the best/most rapid access point to get to the platform
image

### CODE REVIEW

File: source/ptreferential/tests/ptref_test.cpp [hunk 1]
```diff
@@ -154,7 +154,7 @@ BOOST_AUTO_TEST_CASE(get_indexes_test) {
         b.connection("stop3", "stop2", 10 * 60);
     });
 
-    // On cherche à retrouver la ligne 1, en passant le stoparea en filtre
+    // On cherche �� retrouver la ligne 1, en passant le stoparea en filtre
```
Comment (woshilapin) [inline source/ptreferential/tests/ptref_test.cpp]: Mmmm, probably revert the entire file below this line?

File: source/routing/tests/routing_api_test_data.h [hunk 2]
```diff
@@ -282,7 +282,7 @@ struct routing_api_data {
         b.data->geo_ref->ways.push_back(way);
 
         way = new navitia::georef::Way();
-        way->name = "rue ts Entrée/Sortie";  // T->S
+        way->name = "rue ts Entr��e/Sortie";  // T->S
```
Comment (woshilapin) [inline source/routing/tests/routing_api_test_data.h]: Probably revert this.

File: source/ptreferential/tests/ptref_test.cpp (hunk 1 above)
Comment (xlqian) [reply source/ptreferential/tests/ptref_test.cpp]: :man_facepalming:  My editor loves formatting everything to utf-8

File: source/jormungandr/jormungandr/scenarios/helper_classes/transfer.py [hunk 3]
```diff
@@ -131,19 +150,216 @@ def _do_no_access_point_transfer(self, section):
...
 
+        if direct_path and direct_path.journeys:
+            return TransferResult(direct_path, section.origin, section.destination)
+
+        return None
+
     def _aysnc_no_access_point_transfer(self, section):
         return self._future_manager.create_future(self._do_no_access_point_transfer, section)
 
+    def _get_access_points(self, stop_point_uri, predicator=lambda x: x):
```
Comment (pbench) [inline source/jormungandr/jormungandr/scenarios/helper_classes/transfer.py]:
```suggestion
    def _get_access_points(self, stop_point_uri, access_point_filter=lambda x: x):
```

File: source/jormungandr/jormungandr/scenarios/helper_classes/transfer.py [hunk 4]
```diff
@@ -131,19 +150,216 @@ def _do_no_access_point_transfer(self, section):
...
+        """
+        if prev_section_mode in ACCESS_POINTS_PHYSICAL_MODES:
+            return self._get_access_points(
+                section.origin.uri, predicator=lambda access_point: access_point.is_exit
+            )
+        elif next_section_mode in ACCESS_POINTS_PHYSICAL_MODES:
+            return self._get_access_points(
+                section.destination.uri, predicator=lambda access_point: access_point.is_entrance
+            )
+
```
Comment (pbench) [inline source/jormungandr/jormungandr/scenarios/helper_classes/transfer.py:189]:
```suggestion
            )
     else:
           return None

```

File: source/jormungandr/jormungandr/scenarios/helper_classes/transfer.py [hunk 5]
```diff
@@ -131,19 +150,216 @@ def _do_no_access_point_transfer(self, section):
...
+            FallbackModes.walking.name,
+            section.duration * 2,
+            self._request,
+            sub_request_id,
+        )
+
+        # now it's time to find the best combo
+        # (stop_point -> access_points or access_points -> stop_point)
+        best_access_point = self.determinate_the_best_access_point(routing_matrix, access_points)
+
```
Comment (pbench) [inline source/jormungandr/jormungandr/scenarios/helper_classes/transfer.py:287]:
```suggestion
if best_access_point is None:
    return None
```

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py [hunk 6]
```diff
@@ -766,3 +777,34 @@ def complete_transfer(pt_journey, transfer_pool):
         if section.type != response_pb2.TRANSFER:
             continue
         transfer_pool.wait_and_complete(section)
+
+
+def prepend_first_and_append_last_coord(dp, origin, destination):
```
Comment (pbench) [inline source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py]: Maybe split this function in two `prepend_first_coord()` and `append_last_coord()` ?

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py [hunk 7]
```diff
@@ -766,3 +777,34 @@ def complete_transfer(pt_journey, transfer_pool):
...
+    """
+    Insert/append origin and destination's coordinates to the geojson
+    """
+    if not dp or not dp.journeys or not dp.journeys[0].sections:
+        return
+
+    nb_coords = sum((len(sec.street_network.coordinates) for sec in dp.journeys[0].sections))
```
Comment (pbench) [inline source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py:789]: I don't really understand this check. Maybe add some comments ?

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py [hunk 8]
```diff
@@ -766,3 +777,34 @@ def complete_transfer(pt_journey, transfer_pool):
...
+    if nb_coords < 2:
+        return
```
Comment (pbench) [inline source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py]: I don't really understand this check. Maybe add some comments ?

Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py]: it just checks the length of the `geojson` of the street network, there should be at least 2 points in a `geojson` logically

File: source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py (hunk 6 above)
Comment (pbench) [reply source/jormungandr/jormungandr/scenarios/helper_classes/helper_utils.py]: And don't you already have the exact same function [here](https://github.com/hove-io/navitia/blob/fca4622c715418b16686de7a106d836ff1fd4179/source/jormungandr/jormungandr/scenarios/helper_classes/streetnetwork_path.py#L104) ?

**Review by pbench — APPROVED** [review]

**Pull request merged**